torchao/prototype/awq/api.py

```diff
@@ -0,0 +1,133 @@
+    Inserts AWQObserver into Linear layers of a given model.
+
+    Args:
+        model: The model to be modified (in place). Ensure model is on the desired device for calibration
+        n_validation_examples: Number of examples used to validate scale options
+        validation_sequence_len: Number of tokens in each validation example
+        quant_dtype: The data type of the quantized weights. Currently only torch.uint4 is intended to be used but can be used with torch.uint1 -> torch.uint8
+        scale search space size: how many different scale options to try. Original AWQ implementation uses 20. A larger size can lead to better results but takes longer to calibrate
+        group_size: Quantization granularity. Use -1 for channel wise quantization
+    """
+    _is_linear = lambda m, fqn: isinstance(m, torch.nn.Linear)
+    assert quant_dtype in _DTYPE_TO_BIT_WIDTH or quant_dtype == torch.uint8, "Invalid quant_dtype. Please use torch.uint1 .. torch.uint8"
+    # AQT config
+    mapping_type = MappingType.ASYMMETRIC
+    quantization_granularity = PerGroup(group_size)
+    quant_min = 0
+    quant_max = 255 if quant_dtype == torch.uint8 else 2 ** _DTYPE_TO_BIT_WIDTH[quant_dtype] - 1 
+    eps = torch.finfo(torch.float32).eps
+    preserve_zero = True
+    zero_point_dtype = torch.int64
+    zero_point_domain = ZeroPointDomain.INT
+    
+
+    def replace_with_observer(layer):
+        # creates observer and replaces linear layers with AWQObservedLinear layers
+        observer = AWQObserver(
+            layer.weight,
+            layer.bias, 
+            quantization_granularity, 
+            mapping_type,
+            quant_dtype, 
+            n_validation_examples,
+            validation_sequence_len,
+            scale_search_space_size,
+            preserve_zero = preserve_zero,
+            zero_point_domain = zero_point_domain,
+            zero_point_dtype = zero_point_dtype,
+            quant_min=quant_min,
+            quant_max = quant_max,
+            eps = eps)
+        return AWQObservedLinear.from_float(layer, observer)
+    _replace_with_custom_fn_if_matches_filter(model, replace_with_observer, _is_linear)
+
+def _observed_linear_subclass_inserter(constructor):
+    """
+    Replaces unquantized AWQObservedLinear instances with quantized linear instances.
+
+    Args:
+        constructor: the function which applies quantization to the AWQObservedLinear layer
+    """
+    def insert_subclass(observed_linear):
+        # creates the new linear layer using constructor
+        linear = torch.nn.Linear(observed_linear.in_features, observed_linear.out_features, observed_linear.bias!=None, device=observed_linear.weight.device, dtype=observed_linear.weight.dtype)
+        linear.weight = torch.nn.Parameter(constructor(observed_linear), requires_grad=False)
+        linear.bias = observed_linear.bias
+        return linear
+
+    return insert_subclass
+    
+
+def awq_uintx(quant_dtype: torch.dtype = torch.uint4,
```

Comment:
can you just remove `weight_quant_fn` and add `use_hqq` to the function?

Reply:
also when it's uint4, I think it's fine to just use TensorCoreTiledLayout

Reply:
I added that as a feature so maybe people can find ways to compose on top of AWQ esp if they come up with new kernels or what not, but I would agree that its not a necessary feature for an initial release